### DISCUSSION TITLE
Minimum distance for node dragging

### DISCUSSION THREAD

**Discussion by hankportney** [xyflow/xyflow#2712] [New Features]
## Problem

I am trying to implement a text editor within a node, and **cannot sufficiently differentiate clicks on that editor vs drags of the node** because in `useDrag()` ([file link](https://github.com/wbkd/react-flow/blob/769261fead0dbfd11f2c327787b18ffb925fc71f/packages/core/src/hooks/useDrag/index.ts)), when `snapToGrid = false`, drag events kick in for any twitch of mouse movement, even `1px`.  

I've observed clicks having up to `6px` of twitch, but the expected twitch may be greater for some accessibility requirements.

## Proposed solution

**Implement a new prop `dragThreshold`** on `<ReactFlow/>`, which applies the strategy recommended by [this StackOverflow answer](https://stackoverflow.com/a/59741870) to only allow drag after the initial threshold has been exceeded once.  

With this property is non-zero (i.e. `6`), users would need to drag a node beyond that threshold of pixels (i.e. `6px`) to initiate free drag.  A value of `0` in this property would match existing behavior.

This would provide parity with other react dragging libraries that implement a minimum drag distance, like [dnd-kit](https://dndkit.com/) ([storybook example of desired behavior](https://master--5fc05e08a4a65d0021ae0bf2.chromatic.com/?path=/story/core-draggable-hooks-usedraggable--minimum-distance)).

The default value of this prop could be `6`, for accessibility; or `0`, for parity with existing behavior.

## Proposed solution implementation

The following implementation assumes some prop `dragThreshold` has been added to `<ReactFlow>` and is accessible within `useDrag/index.ts`.  **I do not know how to do this with best practices for this library,** lest I would have submitted this as a PR.  Hopefully you can help!

> **Please find attached a [modified `useDrag/index.ts` file](https://github.com/wbkd/react-flow/files/10300554/modifiedUseDrag.txt) (as a `.txt' so it could be attached) with all the below changes to quickly try it out.  That file comes with a hard-coded threshold of 50px.**
---

At the top of `useDrag/index.ts`, inside `useDrag()`...

```jsx
// Convert `dragging` from a `useState` to `useRef` so it survives rerenders, and prevents an issue where the node would re-engage the drag threshold incorrectly during an active drag.
const dragging = useRef<boolean>(false);

  
// Create new variable `startPos` to track start position of the drag.
const startPos = useRef<{ x: number | null; y: number | null }>({
    x: null,
    y: null
});
```

On lines 141 and 154 of `useDrag/index.ts` (multiple locations)...
```jsx
// Replace all `setDragging` state setting to `dragging.current` ref manipulation
dragging.current = true; // On line 141, replace setDragging(true);
dragging.current = false; // On line 154, replace setDragging(false);
```

After line 52 of `useDrag/index.ts`, inside `.on("start", (event: UseDragEvent) => {`...

```jsx
// Record starting position when a drag begins.
startPos.current = getPointerPosition(event);
```

On Line 102 of `useDrag/index.ts`, inside `.on("drag", (event: UseDragEvent) => {`...
```jsx
// Create variable `travelDistance`, the radial distance from `startPos` for this drag event emission.
const xDelta = startPos.current.x != null ? startPos.current.x - pointerPos.x : 0;
const yDelta = startPos.current.y != null ? startPos.current.y - pointerPos.y : 0;
const travelDistance = Math.sqrt(xDelta * xDelta + yDelta * yDelta);

// Create variable `hasExceededThreshold`, a boolean for if we've exceeded threshold now or previously.
// - We've exceeded the drag threshold if we did so during this calculation,
// - but what if we've dragged back to where we started?
// - because `dragging=true` gets set the first time we exceed the threshold,
// - it's safe to assume if dragging is already true, then we've previously exceeded the threshold.
const hasExceededThreshold = travelDistance >= dragThreshold || dragging.current == true; // new prop `dragThreshold` used here

// skip events without movement by snapping rules and events that haven't exceeded the drag threshold.
if (( lastPos.current.x !== pointerPos.xSnapped ||
       lastPos.current.y !== pointerPos.ySnapped) &&
       dragItems.current &&
       hasExceededThreshold // our new function
 ) {
```

**Comment by moklick** [comment]
You need to add the class name "nodrag" to the editor. You should never be able to drag a node within an input field, because you might want to select text. Does that work for you?

**Reply by hankportney** [reply]
Hey Moritz, thanks for the quick reply!

In this case, I actually do want to be able to enable dragging from a text editor, using some situational input differentiation.  Specifically, I am trying to mirror FigJam's sticky notes, where a text editor is superimposed with a drag container, and the editor only triggers under certain input situations.

What FigJam does is *(see attached clip below)*...
- On first **mouse down** on a node, you select it.
  - If you **drag** (> ~6px), you begin dragging the node.
- On second **mouse down** (after you've selected it), nothing happens, but while holding...
  - If you **drag** (> ~6px) while the mouse is down, you begin dragging the node, and don't engage the editor during drag or when releasing the mouse.
  - If you **release the mouse without dragging** (< 6px), your click is conveyed to the editor and your cursor is entered where you let go.
- While in the editor:
  - If you **drag**, you select text, not drag the node
  - You must press `Esc` or click out and re-select to exit the editor and be able to drag again.
  
---
  

I believe the above **solution I recommended is likely only half of what is required:**
1. I know it is needed to ensure the `dragging` property is not incorrectly set to true for twitches during a click when `snapToGrid=false` (when `snapToGrid=true`, this is covered by the grid being a form of drag threshold), which I believe is already valuable for accessibility and parity reasons for the library.
2. I also need some way of ensuring my `contenteditable` element doesn't need `nodrag` to receive input, which I am also having issues with.

What do you think?  Do you know of any other way I could approach this?

---

https://user-images.githubusercontent.com/55362999/209569760-79d9795b-7422-478f-9c11-cf9a8bd26b58.mov

**Reply by jazeee** [reply]
Based on this, it seems like you would conditionally need to add `nodrag` to the block, based on `onFocus` of the input element, or  perhaps `onClick` on some floating element above the input element.
It looks like the figjam video requires you to activate the field to edit it and to drag select inside it.

**Comment by zeltas** [comment]
I find the idea of an input textbox inside a node very interesting as no example is given in custom nodes. Would you mind releasing the textbox code as an tutorial/example?
